service/history/workflow/update/registry.go

```diff
@@ -26,229 +26,202 @@ package update
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
 )
 
 type (
 	Registry interface {
-		Add(request *updatepb.Request) (*Update, Duplicate, RemoveFunc)
+		// FindOrCrate finds an existing Update or creates a new one. The second
+		// return value (bool) indicates whether the Update returned is newly
+		// created (true) for an existing update (false).
+		FindOrCreate(ctx context.Context, protocolInstanceID string) (*Update, bool, error)
 
+		// Find finds an existing update in this Registry but does not create a
+		// new update it is absent.
+		Find(ctx context.Context, protocolInstanceID string) (*Update, bool)
+
+		// CreateOutgoingMessages polls each registered Update for outbound
+		// messages and returns them.
 		CreateOutgoingMessages(startedEventID int64) ([]*protocolpb.Message, error)
 
-		HasPending(filterMessages []*protocolpb.Message) bool
-		ValidateIncomingMessages(messages []*protocolpb.Message) error
-		ProcessIncomingMessages(messages []*protocolpb.Message) error
-	}
+		// HasIncomplete returns true if the registry has Updates that have not
+		// yet completed or are not at least provisionally complete.
+		HasIncomplete() bool
 
-	Duplicate bool
+		// HasOutgoing returns true if the registry has any Updates that want to
+		// sent messages to a worker.
+		HasOutgoing() bool
 
-	RemoveFunc func()
+		// Len observes the number of updates in tis Registry.
+		Len() int
+	}
 
-	// Storage represents the update package's requirements for writing
+	// RegistryStore represents the update package's requirements for writing
 	// events and restoring ephemeral state from an event index.
-	Storage interface {
-		// AddWorkflowExecutionUpdateAcceptedEvent writes an update accepted
-		// event.
-		AddWorkflowExecutionUpdateAcceptedEvent(updateID string, accpt *updatepb.Acceptance) (*historypb.HistoryEvent, error)
-
-		// AddWorkflowExecutionUpdateCompletedEvent writes an update completed
-		// event.
-		AddWorkflowExecutionUpdateCompletedEvent(resp *updatepb.Response) (*historypb.HistoryEvent, error)
-
-		// GetAcceptedWorkflowExecutionUpdateIDs reads from durable state the
-		// set of update IDs that are known to be in the accepted state.
-		GetAcceptedWorkflowExecutionUpdateIDs(ctx context.Context) ([]string, error)
+	RegistryStore interface {
+		GetAcceptedWorkflowExecutionUpdateIDs(context.Context) []string
+		GetUpdateInfo(ctx context.Context, updateID string) (*persistencespb.UpdateInfo, bool)
+		GetUpdateOutcome(ctx context.Context, updateID string) (*updatepb.Outcome, error)
 	}
 
 	RegistryImpl struct {
-		sync.RWMutex
-		updates map[string]*Update
-		store   Storage
+		mu              sync.RWMutex
+		updates         map[string]*Update
+		store           RegistryStore
+		instrumentation instrumentation
 	}
+
+	regOpt func(*RegistryImpl)
 )
 
-var _ Registry = (*RegistryImpl)(nil)
+//revive:disable:unexported-return I *want* it to be unexported
```

Comment:
I don't want external packages to be able to create their own option implementations. I get that there's a hack where it's still possible, but this makes it harder and sends a message.